### PR TITLE
Add GPU support for sparse reorder

### PR DESCRIPTION
**Pull request by kaixih**
This PR adds GPU support to `tf.sparse.reorder`.

cc. @nluehr @benbarsdell

### CODE REVIEW

**Comment by kaixih** [comment]
@sanjoy This PR is related to the sparse_to_dense op. Can you help review? Thanks.

File: tensorflow/core/kernels/sparse_reorder_op_gpu.cu.cc [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+      GpuLaunchConfig config = GetGpuLaunchConfig(num_elems, d);
+      OP_REQUIRES_OK(c, GpuLaunchKernel(
+          IndicesFlattenKernel, config.block_count,
+          config.thread_per_block, 0, d.stream(), indices, num_elems,
+          dims, num_dims, flat_indices));
+
+      Tensor permutes_tensor;
+      OP_REQUIRES_OK(c, c->allocate_temp(DT_INT64, {num_elems},
+                                         &permutes_tensor));
+      auto permutes_data = permutes_tensor.template flat<int64>().data();
```
Comment (sanjoy) [inline tensorflow/core/kernels/sparse_reorder_op_gpu.cu.cc:95]: Maybe s/permutes/permutation/ ?

**Review by sanjoy — CHANGES_REQUESTED** [review]

File: tensorflow/python/kernel_tests/sparse_reorder_op_test.py [hunk 2]
```diff
@@ -57,72 +57,67 @@ def testStaticShapeInfoPreserved(self):
...
-      sp_output = sparse_ops.sparse_reorder(sp_input)
+    sp_input = self._SparseTensorPlaceholder()
+    input_val = self._SparseTensorValue_5x6(np.arange(6))
+    sp_output = sparse_ops.sparse_reorder(sp_input)
 
-      output_val = sess.run(sp_output, {sp_input: input_val})
-      self.assertAllEqual(output_val.indices, input_val.indices)
-      self.assertAllEqual(output_val.values, input_val.values)
-      self.assertAllEqual(output_val.dense_shape, input_val.dense_shape)
+    output_val = sess.run(sp_output, {sp_input: input_val})
```
Comment (sanjoy) [inline tensorflow/python/kernel_tests/sparse_reorder_op_test.py:74]: Where is `sess` defined?

File: tensorflow/core/kernels/sparse_reorder_op_gpu.cu.cc [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    const Eigen::GpuDevice& d = c->eigen_gpu_device();
+
+    const int64 num_elems = input_ind.dims() > 0 ? input_ind.dim_size(0) : 1;
+    const int64 num_dims = input_ind.dims() > 1 ? input_ind.dim_size(1) : 1;
+
+    auto indices = input_ind.template flat<int64>().data();
+    auto values = input_val.template flat<T>().data();
+    auto dims = input_shape_in.template flat<int64>().data();
+
+    if (num_elems <= 0) return;
```
Comment (sanjoy) [inline tensorflow/core/kernels/sparse_reorder_op_gpu.cu.cc:80]: Why not `num_elems == 0`?

**Review by sanjoy — CHANGES_REQUESTED** [review]

File: tensorflow/python/kernel_tests/sparse_reorder_op_test.py (hunk 2 above)
Comment (kaixih) [reply tensorflow/python/kernel_tests/sparse_reorder_op_test.py]: Done.

File: tensorflow/core/kernels/sparse_reorder_op_gpu.cu.cc (hunk 3 above)
Comment (kaixih) [reply tensorflow/core/kernels/sparse_reorder_op_gpu.cu.cc]: Done.

**Review by sanjoy — APPROVED** [review]

**Pull request merged**